network/src/peer_manager/mod.rs

```diff
@@ -560,8 +560,9 @@ where
                 );
                 // Drop the new connection and keep the one already stored in active_peers
                 let drop_fut = async move {
+                    let (_, mut control) = connection.start().await;
```

Comment:
does this work? `let (_, ...) = connection.start().await` will immediately drop the listener `rx` channel, so the next connection event will tear down the driver future, potentially before we get a chance to `.close()`. 

we should probably do `let (_listener, ...) = ` to keep the listener in scope and allow the proper graceful shutdown via the `control.close()`

Reply:
You bring up a very interesting point and I think a better model would be to not drop the connection if sending on the `tx` send of stream listener channel fails. The reason being that a client might simply be uninterested in listening for inbound substreams and only care about making outbound substreams. In such a case, it should be ok to drop the `rx` end of the channel. On the other hand, if we simply exit the loop and drop the connection, this abstraction becomes leaky. Thoughts?